provider/search-provider/src/main/java/dev/ikm/tinkar/provider/search/TypeAheadSearch.java

```diff
@@ -0,0 +1,177 @@
+import org.apache.lucene.search.highlight.Formatter;
+import org.apache.lucene.search.highlight.Highlighter;
+import org.apache.lucene.search.highlight.InvalidTokenOffsetsException;
+import org.apache.lucene.search.highlight.QueryScorer;
+import org.apache.lucene.search.highlight.SimpleHTMLFormatter;
+import org.apache.lucene.search.spell.LuceneDictionary;
+import org.apache.lucene.search.suggest.Lookup;
+import org.apache.lucene.search.suggest.analyzing.AnalyzingSuggester;
+import org.apache.lucene.search.suggest.analyzing.FuzzySuggester;
+import org.eclipse.collections.api.list.ImmutableList;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class TypeAheadSearch {
+    private static final Logger LOG = LoggerFactory.getLogger(TypeAheadSearch.class);
+    public static final String NID = "nid";
+
+    public static final String TEXT_FIELD_NAME = "text";
+
+    private static AnalyzingSuggester suggester;
+    private static FuzzySuggester fuzzySuggester;
+    private static DirectoryReader reader;
+
+    public static void buildSuggester() throws IOException {
+        Stopwatch stopwatch = new Stopwatch();
+        reader = DirectoryReader.open(Indexer.indexWriter);
+        LuceneDictionary dict = new LuceneDictionary(reader, TEXT_FIELD_NAME);
+        suggester = new AnalyzingSuggester(Indexer.indexDirectory, "suggest", Indexer.analyzer);
+        suggester.build(dict);
+        LOG.info("TypeAheadSearch index build duration: {}", stopwatch.durationString());
+    }
+
+    public static void close() throws IOException {
+        if (reader != null) {
+            reader.close();
+        }
+    }
+
+    public static List<String> suggest(String term) throws IOException {
+        List<Lookup.LookupResult> lookup = suggester.lookup(term, false, 100);
+        return lookup.stream().map(a -> a.key.toString()).collect(Collectors.toList());
+    }
+
+    public static void buildFuzzySuggester() throws IOException {
+        Stopwatch stopwatch = new Stopwatch();
+        reader = DirectoryReader.open(Indexer.indexWriter);
+        LuceneDictionary dict = new LuceneDictionary(reader, TEXT_FIELD_NAME);
+        fuzzySuggester = new FuzzySuggester(Indexer.indexDirectory, "suggest", Indexer.analyzer);
+        fuzzySuggester.build(dict);
+        LOG.info("TypeAheadSearch fuzzy index build duration: {}", stopwatch.durationString());
+    }
+
+    public static List<String> fuzzySuggest(String term) throws IOException {
+        List<Lookup.LookupResult> lookup = fuzzySuggester.lookup(term, false, 100);
+        return lookup.stream().map(a -> a.key.toString()).collect(Collectors.toList());
+    }
+
+    public static void buildSearchIndex(String term) throws IOException, ParseException, InvalidTokenOffsetsException {
```

Comment:
This is never called.

Reply:
these are in progress and will be addressed in the next typeahead task.